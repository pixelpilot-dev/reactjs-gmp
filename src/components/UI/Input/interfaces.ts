export interface IInputProps {
  id: string;
  type: 'text' | 'password' | 'tel' | 'email' | 'url';
  placeholder?: string;
  value: string | number | null;
  name: string;
  disabled?: boolean;
  readonly?: boolean;
  className?: string;
}
