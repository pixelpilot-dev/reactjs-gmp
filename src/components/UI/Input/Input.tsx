import * as React from 'react';
import cn from 'classnames';
import { IInputProps } from './interfaces';

import styles from './Input.module.scss';

export const Input: React.FC<IInputProps> = ({
  id,
  type,
  placeholder,
  value,
  name,
  disabled,
  onChange,
  className,
  readonly,
}) => {
  return (
    <input
      id={id}
      type={type}
      name={name}
      value={value}
      placeholder={placeholder}
      disabled={disabled}
      readOnly={readonly}
      onChange={onChange}
      className={cn(styles.input, className)}
    />
  );
};
