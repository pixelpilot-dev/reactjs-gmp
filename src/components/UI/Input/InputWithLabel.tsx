import { makeFieldWithLabel } from '../../hoc/makeFieldWithLabel';
import { Input } from './Input';

export const InputWithLabel = makeFieldWithLabel(Input);
