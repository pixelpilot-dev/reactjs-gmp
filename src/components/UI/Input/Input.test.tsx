import React from 'react';
import '@testing-library/jest-dom/extend-expect';
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { useField } from 'formik';

import { Input } from './Input';
import { InputWithLabel } from './InputWithLabel';

const mockField = {
  name: 'input',
  value: '',
  onBlur: jest.fn(),
  onChange: jest.fn(),
};
const mockMeta = {
  value: '',
  touched: false,
  error: undefined,
  initialValue: '',
  initialTouched: false,
  initialError: false,
};
const mockMetaWithError = {
  touched: true,
  error: 'Required',
  initialError: undefined,
  initialTouched: false,
  initialValue: '',
  value: '',
};
const mockPropsForField = {
  id: 'input',
  placeholder: '',
  disabled: false,
  readonly: false,
  forId: 'input',
  label: 'Input label',
};

jest.mock('formik');

describe('Input component', () => {
  it('renders correctly', () => {
    (useField as jest.Mock).mockReturnValue([mockField, mockMeta]);

    const props = {
      ...mockPropsForField,
      ...mockField,
    };
    const { asFragment } = render(<Input type='text' {...props} />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('should value changed', async () => {
    (useField as jest.Mock).mockReturnValue([mockField, mockMeta]);

    const props = {
      ...mockPropsForField,
      ...mockField,
    };

    const user = userEvent.setup();
    render(<Input type='text' {...props} />);

    await user.type(screen.getByRole('textbox'), 'Test value');
    expect(props.onChange).toBeCalledTimes(10);
  });

  it('should be invalid', () => {
    (useField as jest.Mock).mockReturnValue([mockField, mockMetaWithError]);

    const props = {
      ...mockPropsForField,
      ...mockField,
    };
    const { asFragment } = render(<Input type='text' {...props} />);

    expect(screen.getByRole('textbox')).toHaveClass('isInvalid');
    expect(asFragment()).toMatchSnapshot();
  });

  it('renders with label correctly', () => {
    (useField as jest.Mock).mockReturnValue([mockField, mockMeta]);

    const props = {
      ...mockPropsForField,
      ...mockField,
    };
    const { asFragment } = render(<InputWithLabel type='text' {...props} />);

    expect(screen.getByLabelText('Input label')).toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
  });
});
