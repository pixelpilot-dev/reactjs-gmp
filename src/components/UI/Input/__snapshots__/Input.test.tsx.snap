// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input component renders correctly 1`] = `
<DocumentFragment>
  <input
    class="input"
    id="input"
    name="input"
    placeholder=""
    type="text"
    value=""
  />
</DocumentFragment>
`;

exports[`Input component renders with label correctly 1`] = `
<DocumentFragment>
  <label
    class="label"
    for="input"
  >
    Input label
  </label>
  <input
    class="input"
    id="input"
    name="input"
    placeholder=""
    type="text"
    value=""
  />
</DocumentFragment>
`;

exports[`Input component should be invalid 1`] = `
<DocumentFragment>
  <input
    class="input isInvalid"
    id="input"
    name="input"
    placeholder=""
    type="text"
    value=""
  />
</DocumentFragment>
`;
